### PR TITLE
Faster debugging with watchpoints

### PR DESCRIPTION
**Pull request by chrisc11**

### CODE REVIEW

File: _drafts/cortex-m-watchpoints.md [hunk 1]
```diff
@@ -0,0 +1,1097 @@
...
+It looks like the `len` for the `buf` argument may be incorrect.
+
+```
+(gdb) list s_graphics_buf
+[...]
+40	static uint16_t s_graphics_buf[2];
+[...]
+```
+
+Now we can see the issue! `s_graphics_buf` was defined as a `uint16_t` and we are passing the total size 4. However the function expected the array length, 2. This means we are writing past the end of the buffer into whatever variable comes after it. We can confirm in GDB that `s_registered_handler` has been allocated right after `s_graphics_buf` and is getting corrupted:
```
Comment (franc0is) [inline _drafts/cortex-m-watchpoints.md:388]: Sorry I'm finding this sentence super confusing:

> Now we can see the issue! `s_graphics_buf` was defined as a `uint16_t` and we are passing the total size 4. However the function expected the array length, 2.

isn't array of 2 uint16_t the same as 4 bytes? The wording is confusing, and I'm not sure where the "4" comes from in the code...

File: _drafts/cortex-m-watchpoints.md [hunk 2]
```diff
@@ -0,0 +1,1097 @@
...
+  for (size_t i = 0; i < num_comparators; i++) {
+    sDwtCompCfg *config = &DWT->COMP_CONFIG[i];
+    config->FUNCTION = config->COMP = config->MASK = 0;
+  }
+}
+```
+
+###### Inspecting watchpoints configured with GDB
+
+Let's configure the watchpoints from [above](#dwt-cortex-m-examples) and take trace things!
```
Comment (franc0is) [inline _drafts/cortex-m-watchpoints.md:757]: what do you mean by "take trace things"?

File: _drafts/cortex-m-watchpoints.md [hunk 3]
```diff
@@ -0,0 +1,1097 @@
...
+
+|                         GDB Command <br/> Trigger Condition                         | DWT_FUNCTION | DWT_COMP   | DWT_MASK |
+| :---------------------------------------------------------------------------------: | ------------ | ---------- | -------- |
+|             `watch *(uint8_t *)0x20000001` <br /> (Write to 0x20000001)             | 0x6          | 0x20000001 | 0x0      |
+|  `watch *(uint16_t *)0x20000000` <br /> (Any write between 0x20000000-0x20000002 )  | 0x6          | 0x20000000 | 0x1      |
+| `watch *(uint8_t[4] *)0x20000010` <br /> (Any write between 0x20000010- 0x20000014) | 0x6          | 0x20000010 | 0x2      |
+| `rwatch *(uint8_t[16] *)0x20000020` <br /> (Any read between 0x20000020-0x2000002F) | 0x5          | 0x20000020 | 0x4      |
+
+> Note: An important observation here is it is only possible to watch ranges of addresses that are aligned on the size being watched. For example, if you wanted to watch 4 bytes starting at 0x20000002, you would need two watchpoints (`DWT_COMP=0x20000002`, `DWT_MASK=1` & `DWT_COMP=0x20000004`, `DWT_MASK=1`).
```
Comment (franc0is) [inline _drafts/cortex-m-watchpoints.md:686]: Worth mentioning whether or not the debugger manages that complexity for you

File: _drafts/cortex-m-watchpoints.md [hunk 4]
```diff
@@ -0,0 +1,1097 @@
...
+
+  sDwtCompCfg *config = &DWT->COMP_CONFIG[comp_id];
+  config->COMP = comp;
+  config->MASK = mask;
+  // set list since this will enable the comparator
+  config->FUNCTION = func;
+}
+```
+
+For example, let's we can enable one over the first 16 bytes of the `g_array` at address `0x200029a0` that we were examining earlier:
```
Comment (franc0is) [inline _drafts/cortex-m-watchpoints.md:838]:
```suggestion
For example, let's enable one over the first 16 bytes of the `g_array` at address `0x200029a0` that we were examining earlier:
```

File: _drafts/cortex-m-watchpoints.md [hunk 5]
```diff
@@ -0,0 +1,1097 @@
...
+
+For the most part, things look like they line up with expectations. For the read watchpoint at `0x20000020`, we expect to see `0x00000006` for `DWT_FUNC1` but some extra bits were left set. These bits map to `DATAVSIZE` which is not used in our current config and `LNK1ENA` which is a read only bit so these settings are fine.
+
+What is more problematic is the `DWT_MASK2` setting. It should be `4` to watch the 16 bytes that was requested. However, it was set to 2. It appears there is a limitation of the SEGGER watchpoint implementation and only 1, 2, and 4-byte watchpoints work.
+
+## Advanced Topics
+
+### Configuring Watchpoints in C Code
+
+It's also possible to generate watchpoints from C code:
```
Comment (franc0is) [inline _drafts/cortex-m-watchpoints.md:819]:
```suggestion
It's also possible to generate watchpoints from C code. Here's a function that does just that:
```

File: _drafts/cortex-m-watchpoints.md [hunk 6]
```diff
@@ -0,0 +1,1097 @@
...
+# Access at offset 16 should not halt since we are only watching the first 16 bytes
+shell> arr_write 16 1
+Write - Addr: 0x200029b0, Index: 16, Value: 0x00000001
+
+# Access to an array we are watch
+shell> arr_write 15 2
+Write - Addr: 0x200029af, Index: 15, Value: 0x00000002
+# Debugger should have halted
+```
+
```
Comment (franc0is) [inline _drafts/cortex-m-watchpoints.md:851]: Add a line here "as expected, our debugger halts:"

File: _drafts/cortex-m-watchpoints.md [hunk 7]
```diff
@@ -0,0 +1,1097 @@
...
+            print("Setting Address Watchpoint")
+            Dwt.set_addr_watchpoint(comp, size)
+        else:
+            print("Setting Data Value Watchpoint")
+            Dwt.set_data_value_watchpoint(comp, size)
+
+
+WatchpointExtensionCliCmd()
+```
+
```
Comment (franc0is) [inline _drafts/cortex-m-watchpoints.md:1011]: Maybe add a brief summary of the features implemented as a list of bullet points?

Comment (franc0is) [reply _drafts/cortex-m-watchpoints.md:1011]: Will watchpoints set with your script register with GDB as standard watchpoints? Can they be disabled with GDB commands or is `watchpoint_ext` required to disable them?

File: _drafts/cortex-m-watchpoints.md [hunk 8]
```diff
@@ -0,0 +1,1097 @@
...
+{% include submit-pr.html %}
+
+{:.no_toc}
+
+## Further Reading
+
+I haven't found too many resources about watchpoints on the web, but here's a few other articles on the topic I've enjoyed:
+
+- [Making use of Cortex-M watchpoints at Runtime](https://m0agx.eu/2018/08/25/cortex-m-debugging-runtime-memory-corruption/)
```
Comment (tyhoff) [inline _drafts/cortex-m-watchpoints.md:1081]: This kind of does. https://www.embeddedrelated.com/showarticle/1330.php

File: _drafts/cortex-m-watchpoints.md [hunk 9]
```diff
@@ -0,0 +1,1097 @@
+---
```
Comment (tyhoff) [inline _drafts/cortex-m-watchpoints.md:1]: Maybe the dumbest example you are missing is something like:

```
for (int i = 0; i < 1000; i++) {
    // I want to do something when i = 100
}
```

I've seen people click the space bar 100 times. I've been to a conference talk where they did this. Might be worth mentioning directly.

File: _drafts/cortex-m-watchpoints.md [hunk 10]
```diff
@@ -0,0 +1,1097 @@
...
+
+Old value = "\000"
+New value = "\001"
+0x000008f6 in prv_arr_write () at ./src/shell_commands.c:153
+153   return 0;
+```
+
+##### Installing a Read Watchpoint
+
+We can also add a watchpoint that will only trigger on a read access using `rwatch`. Let's try it out:
```
[the remark this answers is not in the document]
Comment (chrisc11) [reply _drafts/cortex-m-watchpoints.md:201]: 👍 reworked the section titles a bit

File: _drafts/cortex-m-watchpoints.md (hunk 1 above)
Comment (chrisc11) [reply _drafts/cortex-m-watchpoints.md]: I've tried to rework it ... the bug I was using as an example here was a situation where ARRAY_SIZE() was expected but sizeof() was passed instead.

File: _drafts/cortex-m-watchpoints.md (hunk 7 above)
Comment (chrisc11) [reply _drafts/cortex-m-watchpoints.md:1050]: Done! The script is definitely just a starting point / POC so I tried to call that out.

**Review by franc0is — APPROVED** [review]
I think your title is pretty good, but in case you want more suggestions:
- "Catch stack overflows and memory corruption with watchpoints"
- "Fixing data corruptions with watchpoints"

**Pull request merged**